README.md
<!--
SPDX-FileCopyrightText: 2013 go-bench AUTHORS
SPDX-License-Identifier: CC-BY-4.0
-->
<!-- markdownlint-disable MD041-->
[![Go Report Card](https://goreportcard.com/badge/github.com/EricNeid/go-bench?style=flat-square)](https://goreportcard.com/report/github.com/EricNeid/go-getdockerimage)
![Go](https://github.com/EricNeid/go-bench/workflows/Go/badge.svg)
[![Release](https://img.shields.io/github/release/EricNeid/go-bench.svg?style=flat-square)](https://github.com/EricNeid/go-bench/releases/latest)
[![Gitpod Ready-to-Code](https://img.shields.io/badge/Gitpod-Ready--to--Code-blue?logo=gitpod)](https://gitpod.io/#https://github.com/EricNeid/go-bench)

# About

This tool is a simple benchmark tool to test the performance and throughput of your server.

Forked from <https://github.com/cmpxchg16/gobench>, so be sure to check
their project out as well.

## Quickstart

Checkout the project and run make (given that go build chain is installed.
Hint: You can simply click on the Gitpod link above and compile it online without setting up a tool chain.

```bash
make all
```

## Installation

Simply grep the latest available binary from the release page and put it somewhere in your path.

If go is installed and $GOPATH/bin is in your path, you can download and install the tool directly
by using go install.

Download tool:

```bash
go install github.com/EricNeid/go-bench/cmd/gobench@v0.3.0
```

## Usage

Running HTTP GET:

```bash
gobench -u http://localhost:80 -k=true -c 500 -t 10
```

Running HTTP Post:

```bash
gobench -u http://localhost:80 -k=true -c 500 -t 10 -d ./data.json
gobench -u http://localhost:80 -k=true -c 500 -t 10 -b '{\"name\":\"Timmy\"}'
```

Getting help:

```bash
gobench --help
```

## Question or comments

Please feel free to open a new issue:
<https://github.com/EricNeid/go-bench/issues>


CHANGELOG.md
<!--
SPDX-FileCopyrightText: 2021 Eric Neidhardt
SPDX-License-Identifier: CC-BY-4.0
-->
<!-- markdownlint-disable MD022 MD032 MD024-->
# Changelog

All notable changes to this project will be documented in this file.

## Unreleased

### Changed (Breaking)
* Switched from fasthttp to net/http -> statistics from v0.2.0 are not comparable
### Added
* display version when printing usage
* Setting for additional headers
* Changelog.md

## 0.2.0 - 2021-08-12
### Added
* Precompiled binaries

## 0.1.0 - 2021-07-23
### Added
* Initial release


client/client_test.go
// SPDX-FileCopyrightText: 2021 Eric Neidhardt
// SPDX-License-Identifier: MIT
package client

import (
	"context"
	"io"
	"net/http"
	"testing"
	"time"

	"github.com/EricNeid/go-bench/internal/verify"
)

func TestNewRequest(t *testing.T) {
	// action
	result := NewRequest("http://localhost", "", "{\"test\":\"value\"}", "application/json", true, "authorizationHeader", "key1=value1,key2=value2, key3=value3")
	// verify
	verify.NotNil(t, result, "Request is nil")
	verify.Equals(t, "http://localhost", result.Url)
	verify.Equals(t, []byte("{\"test\":\"value\"}"), result.PostBody)
	verify.Equals(t, "application/json", result.ContentType)
	verify.Equals(t, true, result.KeepAlive)
	verify.Equals(t, 4, len(result.AdditionalHeaders))
	verify.Equals(t, "authorizationHeader", result.AdditionalHeaders["Authorization"])
	verify.Equals(t, "value1", result.AdditionalHeaders["key1"])
	verify.Equals(t, "value2", result.AdditionalHeaders["key2"])
	verify.Equals(t, "value3", result.AdditionalHeaders["key3"])
}

func TestPerformRequest_get(t *testing.T) {
	// arrange
	requestReceived := false
	// create test server
	handler := http.NewServeMux()
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		requestReceived = true
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("test response"))
	})
	srv := http.Server{Addr: "localhost:8080", Handler: handler}
	go srv.ListenAndServe()
	defer srv.Close()
	// create unit
	unit := Client{
		Request: Request{
			Url: "http://localhost:8080/",
		},
	}
	// action
	unit.PerformRequest()
	// verify
	verify.Assert(t, requestReceived, "Request was not received")
	verify.Equals(t, 1, unit.Statistic.RequestCount)
	verify.Equals(t, 1, unit.Statistic.SuccessCount)
	verify.Equals(t, 0, unit.Statistic.FailureCount)
	verify.Equals(t, 0, unit.Statistic.NetworkFailedCount)
	verify.Equals(t, 0, unit.Statistic.IOFailedCount)
	verify.Equals(t, int64(len([]byte("test response"))), unit.Statistic.ReadThroughput)
	verify.Equals(t, int64(0), unit.Statistic.WriteThroughput)
}

func TestPerformRequestWithContext_shouldCancelRequestAfterDeadline(t *testing.T) {
	// arrange
	// create test server
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(1)*time.Second)
	defer cancel()
	handler := http.NewServeMux()
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		// let client hang until timeout
		<-ctx.Done()
	})
	srv := http.Server{Addr: "localhost:8080", Handler: handler}
	go srv.ListenAndServe()
	defer srv.Close()
	// create unit
	unit := Client{
		Request: Request{
			Url: "http://localhost:8080/",
		},
	}
	// action
	unit.PerformRequestWithContent(ctx)
	// verify
	verify.Equals(t, 1, unit.Statistic.RequestCount)
	verify.Equals(t, 0, unit.Statistic.FailureCount)
	verify.Equals(t, 1, unit.Statistic.NetworkFailedCount)
	verify.Equals(t, 0, unit.Statistic.IOFailedCount)
}

func TestPerformRequest_withCustomHeader(t *testing.T) {
	// arrange
	requestReceived := false
	// create test server
	handler := http.NewServeMux()
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("custom-header-field") == "test header value" {
			requestReceived = true
		}
		w.WriteHeader(http.StatusOK)
	})
	srv := http.Server{Addr: "localhost:8080", Handler: handler}
	go srv.ListenAndServe()
	defer srv.Close()
	// create unit
	unit := Client{
		Request: Request{
			Url: "http://localhost:8080/",
			AdditionalHeaders: map[string]string{
				"custom-header-field": "test header value",
			},
		},
	}
	// action
	unit.PerformRequest()
	// verify
	verify.Assert(t, requestReceived, "Request was not received")
}

func TestPerformRequest_post(t *testing.T) {
	// arrange
	requestReceived := false
	// create test server
	handler := http.NewServeMux()
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if body, err := io.ReadAll(r.Body); err == nil && string(body) == "test body" {
			requestReceived = true
		}
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("test response"))
	})
	srv := http.Server{Addr: "localhost:8080", Handler: handler}
	go srv.ListenAndServe()
	defer srv.Close()
	// create unit
	unit := Client{
		Request: Request{
			Url:         "http://localhost:8080/",
			PostBody:    []byte("test body"),
			ContentType: "text/string",
		},
	}
	// action
	unit.PerformRequest()
	// verify
	verify.Assert(t, requestReceived, "Request was not received")
	verify.Equals(t, 1, unit.Statistic.RequestCount)
	verify.Equals(t, 1, unit.Statistic.SuccessCount)
	verify.Equals(t, 0, unit.Statistic.FailureCount)
	verify.Equals(t, 0, unit.Statistic.NetworkFailedCount)
	verify.Equals(t, 0, unit.Statistic.IOFailedCount)
	verify.Equals(t, int64(len([]byte("test response"))), unit.Statistic.ReadThroughput)
	verify.Equals(t, int64(len("test body")), unit.Statistic.WriteThroughput)
}

func TestRunForAmount(t *testing.T) {
	// arrange
	receivedCount := 0
	// create test server
	handler := http.NewServeMux()
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		receivedCount++
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("test response"))
	})
	srv := http.Server{Addr: "localhost:8080", Handler: handler}
	go srv.ListenAndServe()
	defer srv.Close()
	// create unit
	unit := Client{
		Request: Request{
			Url: "http://localhost:8080/",
		},
	}
	// action
	unit.RunForAmount(10)
	// verify
	verify.Equals(t, 10, receivedCount)
	verify.Equals(t, 10, unit.Statistic.RequestCount)
	verify.Equals(t, 10, unit.Statistic.SuccessCount)
	verify.Equals(t, 0, unit.Statistic.FailureCount)
	verify.Equals(t, 0, unit.Statistic.NetworkFailedCount)
	verify.Equals(t, 0, unit.Statistic.IOFailedCount)
	verify.Equals(t, 10*int64(len([]byte("test response"))), unit.Statistic.ReadThroughput)
}

func TestRunForDuration(t *testing.T) {
	// arrange
	receivedCount := 0
	// create test server
	handler := http.NewServeMux()
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("test response"))
		receivedCount++
	})
	srv := http.Server{Addr: "localhost:8080", Handler: handler}
	go srv.ListenAndServe()
	defer srv.Close()
	// create unit
	unit := Client{
		Request: Request{
			Url: "http://localhost:8080/",
		},
	}
	// action
	unit.RunForDuration(1 * time.Second)
	// verify
	verify.Assert(t, receivedCount > 0, "No request received")
	verify.Assert(t, unit.Statistic.RequestCount > 0, "No request received")
	verify.Equals(t, unit.Statistic.RequestCount, unit.Statistic.SuccessCount)
	verify.Equals(t, int64(unit.Statistic.SuccessCount*len([]byte("test response"))), unit.Statistic.ReadThroughput)
}


client/client.go
// SPDX-FileCopyrightText: 2021 Eric Neidhardt
// SPDX-License-Identifier: MIT
package client

import (
	"bytes"
	"context"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

type Request struct {
	Url string

	PostBody    []byte
	ContentType string

	KeepAlive         bool
	AdditionalHeaders map[string]string
}

type Statistic struct {
	// Overall number of bytes read.
	ReadThroughput int64
	// Overall number of bytes written.
	WriteThroughput int64

	// Overall number of performed requests, always equal to the sum of failed and successfull requests.
	RequestCount int
	SuccessCount int
	// Number of requests that failed with status code != 2xx.
	FailureCount int
	// Number of request that failed with error != nil while performing the request.
	NetworkFailedCount int
	// Number of request that failed with error != nil while reading response.
	IOFailedCount int
}

type Client struct {
	Statistic  Statistic
	Request    Request
	HttpClient http.Client
}

func NewRequest(
	url string,
	postDataFilePath string,
	postBody string,
	contentType string,
	keepAlive bool,
	authHeader string,
	additionalHeaders string,
) *Request {
	// preparing request
	var request = Request{
		Url:       url,
		KeepAlive: keepAlive,
	}

	// read optional post body
	if postDataFilePath != "" {
		data, err := os.ReadFile(postDataFilePath)
		if err != nil {
			log.Fatalf("Error while reading post body from file: %s Error: %s", postDataFilePath, err)
		}
		request.PostBody = data
	} else if postBody != "" {
		request.PostBody = []byte(postBody)
	}

	// create headers
	request.AdditionalHeaders = make(map[string]string)
	if contentType != "" {
		request.ContentType = contentType
	}
	if authHeader != "" {
		request.AdditionalHeaders["Authorization"] = authHeader
	}
	headers := strings.Split(additionalHeaders, ",")
	for _, header := range headers {
		keyValue := strings.Split(header, "=")
		if len(keyValue) == 2 {
			request.AdditionalHeaders[strings.TrimSpace(keyValue[0])] = strings.TrimSpace(keyValue[1])
		}
	}

	return &request
}

func NewClient(timeout time.Duration, Request Request) *Client {
	return &Client{
		HttpClient: http.Client{
			Timeout: timeout,
		},
		Request: Request,
	}
}

func (c *Client) RunForDuration(timeout time.Duration) {
	startTime := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	for time.Since(startTime) < timeout {
		c.PerformRequestWithContent(ctx)
	}
	// the last request can be interrupted by context timeout
	// we remove this from statistic
	if c.Statistic.NetworkFailedCount == 1 {
		c.Statistic.RequestCount--
		c.Statistic.NetworkFailedCount--
	}
}

func (c *Client) RunForAmount(requestCount int) {
	for i := 0; i < requestCount; i++ {
		c.PerformRequest()
	}
}

func (c *Client) PerformRequest() {
	c.PerformRequestWithContent(context.Background())
}

func (c *Client) PerformRequestWithContent(ctx context.Context) {
	// prepare request from configuration
	var req *http.Request
	var err error
	if c.Request.PostBody != nil {
		req, err = http.NewRequestWithContext(ctx, "POST", c.Request.Url, bytes.NewReader(c.Request.PostBody))
		req.Header.Set("Content-Type", c.Request.ContentType)
	} else {
		req, err = http.NewRequestWithContext(ctx, "GET", c.Request.Url, nil)
	}
	if err != nil {
		panic("Could not create http request")
	}

	if c.Request.KeepAlive {
		req.Header.Set("Connection", "keep-alive")
	} else {
		req.Header.Set("Connection", "close")
	}

	if c.Request.AdditionalHeaders != nil {
		for k, v := range c.Request.AdditionalHeaders {
			req.Header.Set(k, v)
		}
	}

	// perform request
	c.Statistic.RequestCount++
	resp, err := c.HttpClient.Do(req)
	if err != nil {
		c.Statistic.NetworkFailedCount++
		return
	}
	defer resp.Body.Close()

	// write statistic
	switch {
	case resp.StatusCode >= 200 && resp.StatusCode <= 299:
		c.Statistic.SuccessCount++
	default:
		c.Statistic.FailureCount++
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		c.Statistic.IOFailedCount++
	}
	c.Statistic.ReadThroughput += int64(len(body))
	c.Statistic.WriteThroughput += int64(len(c.Request.PostBody))
}


cmd/gobench/main.go
// SPDX-FileCopyrightText: 2021 Eric Neidhardt
// SPDX-License-Identifier: MIT
package main

import (
	"flag"
	"fmt"
	"os"
	"sync"
	"time"

	"github.com/EricNeid/go-bench/client"
)

const version = "0.3.0"

var (
	clientCount int = 100

	requestCount        int = -1
	requestsDurationSec int = -1

	url string = ""

	postDataFilePath string = ""
	postBody         string = ""
	contentType      string = ""

	keepAlive bool = false

	clientTimeoutMs int64 = 10 * 1000 // 10 seconds

	authHeader        string = ""
	additionalHeaders string = ""
)

func init() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
		fmt.Printf("Version: %s\n", version)
		fmt.Printf("Command line options:\n")
		flag.PrintDefaults()
	}
	flag.IntVar(&clientCount, "c", clientCount, "Number of concurrent clients")
	flag.IntVar(&requestCount, "r", requestCount, "Number of requests per client")
	flag.IntVar(&requestsDurationSec, "t", requestsDurationSec, "Duration for performing requests (in seconds)")

	flag.StringVar(&url, "u", url, "URL")

	flag.StringVar(&postDataFilePath, "d", postDataFilePath, "HTTP POST data file path: gobench -u http://localhost -t 10 -d ./data.json")
	flag.StringVar(&postBody, "b", postBody, "HTTP POST body: gobench -u http://localhost -t 10 -b '{\"name\":\"max\"}'")
	flag.StringVar(&contentType, "content-type", contentType, "Content type of post body")

	flag.BoolVar(&keepAlive, "k", keepAlive, "Do HTTP keep-alive ")
	flag.Int64Var(&clientTimeoutMs, "timeout", clientTimeoutMs, "Timeout (in milliseconds)")

	flag.StringVar(&authHeader, "auth", authHeader, "Authorization header: gobench -u http://localhost -t 10 -auth 'Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ=='")
	flag.StringVar(
		&additionalHeaders, "headers", additionalHeaders, "additional header fields: gobench -u http://localhost -t 10 -headers key1=value1,key2=value2",
	)

	flag.Parse()

	if url == "" {
		println("Url is required")
		flag.Usage()
		os.Exit(1)
	}

	if requestCount == -1 && requestsDurationSec == -1 {
		fmt.Println("Request count or request duration must be provided")
		flag.Usage()
		os.Exit(1)
	}

	if requestCount != -1 && requestsDurationSec != -1 {
		fmt.Println("Only one should be provided: [requests|duration]")
		flag.Usage()
		os.Exit(1)
	}

	if clientCount <= 0 {
		fmt.Println("Number of clients must be larger than 0")
		flag.Usage()
		os.Exit(1)
	}
}

func main() {
	request := client.NewRequest(url, postDataFilePath, postBody, contentType, keepAlive, authHeader, additionalHeaders)

	var clients []*client.Client
	for i := 0; i < clientCount; i++ {
		clients = append(clients, client.NewClient(time.Duration(clientTimeoutMs)*time.Millisecond, *request))
	}

	fmt.Printf("Dispatching %d clients\n", len(clients))

	var done sync.WaitGroup
	done.Add(len(clients))
	startTime := time.Now()

	if requestCount != -1 {
		for _, c := range clients {
			go func(c *client.Client) {
				c.RunForAmount(requestCount)
				done.Done()
			}(c)
		}
	} else if requestsDurationSec != -1 {
		for _, c := range clients {
			go func(c *client.Client) {
				c.RunForDuration(time.Duration(requestsDurationSec) * time.Second)
				done.Done()
			}(c)
		}
	}
	fmt.Println("Waiting for results...")
	done.Wait()

	printResults(clients, startTime)
}

func printResults(clients []*client.Client, startTime time.Time) {
	var requests int64
	var success int64
	var failed int64
	var networkFailed int64

	var readThroughput int64
	var writeThroughput int64

	for _, c := range clients {
		requests += int64(c.Statistic.RequestCount)
		success += int64(c.Statistic.SuccessCount)
		networkFailed += int64(c.Statistic.NetworkFailedCount)
		failed += int64(c.Statistic.FailureCount)
		readThroughput += int64(c.Statistic.ReadThroughput)
		writeThroughput += int64(c.Statistic.WriteThroughput)
	}

	elapsed := int64(time.Since(startTime).Seconds())

	if elapsed == 0 {
		elapsed = 1
	}

	fmt.Println()
	fmt.Printf("Requests:                       %10d hits\n", requests)
	fmt.Printf("Successful requests:            %10d hits\n", success)
	fmt.Printf("Network failed:                 %10d hits\n", networkFailed)
	fmt.Printf("Bad requests failed (!2xx):     %10d hits\n", failed)
	fmt.Printf("Successful requests rate:       %10d hits/sec\n", success/elapsed)
	fmt.Printf("Read throughput:                %10d bytes/sec\n", readThroughput/elapsed)
	fmt.Printf("Write throughput:               %10d bytes/sec\n", writeThroughput/elapsed)
	fmt.Printf("Test time:                      %10d sec\n", elapsed)
}


cmd/gobench/dialer.go
// SPDX-FileCopyrightText: 2013 go-bench AUTHORS
// SPDX-License-Identifier: BSD-3-Clause
package main

import (
	"net"
	"sync"
	"sync/atomic"

	"github.com/valyala/fasthttp"
)

type Result struct {
	requests      int64
	success       int64
	networkFailed int64
	badFailed     int64
}

type MyConn struct {
	net.Conn
}

var readThroughput int64
var writeThroughput int64

// override fasthttp Read function
func (conn *MyConn) Read(b []byte) (n int, err error) {
	len, err := conn.Conn.Read(b)

	if err == nil {
		atomic.AddInt64(&readThroughput, int64(len))
	}

	return len, err
}

// override fasthttp Write function
func (conn *MyConn) Write(b []byte) (n int, err error) {
	len, err := conn.Conn.Write(b)

	if err == nil {
		atomic.AddInt64(&writeThroughput, int64(len))
	}

	return len, err
}

func newMyDialFunction() func(address string) (conn net.Conn, err error) {
	return func(address string) (net.Conn, error) {
		conn, err := net.Dial("tcp", address)
		if err != nil {
			return nil, err
		}

		myConn := &MyConn{Conn: conn}

		return myConn, nil
	}
}

func startClient(configuration runConfiguration, result *Result, done *sync.WaitGroup) {
	// either perform requests until request count is reached or wait for timeout to kick in
	for result.requests < configuration.requestCount || configuration.requestsDuration != -1 {
		for _, tmpUrl := range configuration.urls {

			req := fasthttp.AcquireRequest()

			req.SetRequestURI(tmpUrl)
			req.Header.SetMethodBytes([]byte(configuration.method))

			if configuration.keepAlive {
				req.Header.Set("Connection", "keep-alive")
			} else {
				req.Header.Set("Connection", "close")
			}

			if configuration.contentType != "" {
				req.Header.SetContentType(configuration.contentType)
			}

			for k, v := range configuration.additioanlHeadrs {
				req.Header.Set(k, v)
			}

			req.SetBody(configuration.postData)

			resp := fasthttp.AcquireResponse()
			err := configuration.myClient.Do(req, resp)
			statusCode := resp.StatusCode()
			result.requests++
			fasthttp.ReleaseRequest(req)
			fasthttp.ReleaseResponse(resp)

			if err != nil {
				result.networkFailed++
				continue
			}

			// check for any success status code
			if statusCode >= 200 && statusCode <= 226 {
				result.success++
			} else {
				result.badFailed++
			}
		}
	}

	done.Done()
}


internal/verify/testing_utils.go
// SPDX-FileCopyrightText: 2021 Eric Neidhardt
// SPDX-License-Identifier: MIT
package verify

import (
	"fmt"
	"path/filepath"
	"reflect"
	"runtime"
	"testing"
)

// Assert fails the test if the condition is false.
func Assert(t *testing.T, condition bool, msg string, v ...interface{}) {
	if !condition {
		_, file, line, _ := runtime.Caller(1)
		fmt.Printf("%s:%d: "+msg+"\n\n", append([]interface{}{filepath.Base(file), line}, v...)...)
		t.FailNow()
	}
}

// Ok fails the test if an err is not nil.
func Ok(t *testing.T, err error) {
	if err != nil {
		_, file, line, _ := runtime.Caller(1)
		fmt.Printf("%s:%d: unexpected error: %s\n\n", filepath.Base(file), line, err.Error())
		t.FailNow()
	}
}

// Equals fails the test if exp is not equal to act.
func Equals(t *testing.T, exp, act interface{}) {
	if !reflect.DeepEqual(exp, act) {
		_, file, line, _ := runtime.Caller(1)
		fmt.Printf("%s:%d:\n\n\texp: %#v\n\n\tgot: %#v\n\n", filepath.Base(file), line, exp, act)
		t.FailNow()
	}
}

// NotNil fails if act is nil
func NotNil(t *testing.T, act interface{}, msg string) {
	if act == nil {
		_, file, line, _ := runtime.Caller(1)
		fmt.Printf("%s:%d: "+msg+"\n\n", append([]interface{}{filepath.Base(file), line}))
		t.FailNow()
	}
}


cmd/gobench/gobench.go
package main

import (
	"bufio"
	"bytes"
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"os/signal"
	"runtime"
	"sync"
	"sync/atomic"
	"time"

	"github.com/valyala/fasthttp"
)

var (
	requestCount     int64
	requestsDuration int64
	clients          int
	url              string
	urlsFilePath     string
	keepAlive        bool
	postDataFilePath string
	postBody         string
	contentType      string
	writeTimeout     int
	readTimeout      int
	authHeader       string
)

type configuration struct {
	urls             []string
	method           string
	postData         []byte
	contentType      string
	requestCount     int64
	requestsDuration int64
	keepAlive        bool
	authHeader       string

	myClient fasthttp.Client
}

type Result struct {
	requests      int64
	success       int64
	networkFailed int64
	badFailed     int64
}

type MyConn struct {
	net.Conn
}

var readThroughput int64
var writeThroughput int64

// override fasthttp Read function
func (conn *MyConn) Read(b []byte) (n int, err error) {
	len, err := conn.Conn.Read(b)

	if err == nil {
		atomic.AddInt64(&readThroughput, int64(len))
	}

	return len, err
}

// override fasthttp Write function
func (conn *MyConn) Write(b []byte) (n int, err error) {
	len, err := conn.Conn.Write(b)

	if err == nil {
		atomic.AddInt64(&writeThroughput, int64(len))
	}

	return len, err
}

func init() {
	flag.Int64Var(&requestCount, "r", -1, "Number of requests per client")
	flag.IntVar(&clients, "c", 100, "Number of concurrent clients")
	flag.StringVar(&url, "u", "", "URL")
	flag.StringVar(&urlsFilePath, "f", "", "URL's file path (line seperated)")
	flag.BoolVar(&keepAlive, "k", true, "Do HTTP keep-alive ")
	flag.StringVar(&postDataFilePath, "d", "", "HTTP POST data file path: gobench -u http://localhost -t 10 -d ./data.json")
	flag.StringVar(&postBody, "b", "", "HTTP POST body: gobench -u http://localhost -t 10 -b '{\"name\":\"max\"}'")
	flag.StringVar(&contentType, "content-type", "", "Content type of post body")
	flag.Int64Var(&requestsDuration, "t", -1, "Period of time (in seconds)")
	flag.IntVar(&writeTimeout, "tw", 5000, "Write timeout (in milliseconds)")
	flag.IntVar(&readTimeout, "tr", 5000, "Read timeout (in milliseconds)")
	flag.StringVar(&authHeader, "auth", "", "Authorization header: gobench -u http://localhost -t 10 -auth 'Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ=='")
}

func printResults(results map[int]*Result, startTime time.Time) {
	var requests int64
	var success int64
	var networkFailed int64
	var badFailed int64

	for _, result := range results {
		requests += result.requests
		success += result.success
		networkFailed += result.networkFailed
		badFailed += result.badFailed
	}

	elapsed := int64(time.Since(startTime).Seconds())

	if elapsed == 0 {
		elapsed = 1
	}

	fmt.Println()
	fmt.Printf("Requests:                       %10d hits\n", requests)
	fmt.Printf("Successful requests:            %10d hits\n", success)
	fmt.Printf("Network failed:                 %10d hits\n", networkFailed)
	fmt.Printf("Bad requests failed (!2xx):     %10d hits\n", badFailed)
	fmt.Printf("Successful requests rate:       %10d hits/sec\n", success/elapsed)
	fmt.Printf("Read throughput:                %10d bytes/sec\n", readThroughput/elapsed)
	fmt.Printf("Write throughput:               %10d bytes/sec\n", writeThroughput/elapsed)
	fmt.Printf("Test time:                      %10d sec\n", elapsed)
}

func readLines(path string) (lines []string, err error) {
	var file *os.File
	var part []byte
	var prefix bool

	if file, err = os.Open(path); err != nil {
		return
	}
	defer file.Close()

	reader := bufio.NewReader(file)
	buffer := bytes.NewBuffer(make([]byte, 0))
	for {
		if part, prefix, err = reader.ReadLine(); err != nil {
			break
		}
		buffer.Write(part)
		if !prefix {
			lines = append(lines, buffer.String())
			buffer.Reset()
		}
	}
	if err == io.EOF {
		err = nil
	}
	return
}

func newConfiguration() configuration {
	flag.Parse()
	if urlsFilePath == "" && url == "" {
		flag.Usage()
		os.Exit(1)
	}

	if requestCount == -1 && requestsDuration == -1 {
		fmt.Println("Requests or period must be provided")
		flag.Usage()
		os.Exit(1)
	}

	if requestCount != -1 && requestsDuration != -1 {
		fmt.Println("Only one should be provided: [requests|period]")
		flag.Usage()
		os.Exit(1)
	}

	configuration := configuration{
		urls:             make([]string, 0),
		method:           "GET",
		postData:         nil,
		keepAlive:        keepAlive,
		authHeader:       authHeader,
		requestCount:     requestCount,
		requestsDuration: requestsDuration,
	}

	if urlsFilePath != "" {
		fileLines, err := readLines(urlsFilePath)

		if err != nil {
			log.Fatalf("Error in ioutil.ReadFile for file: %s Error: %v", urlsFilePath, err)
		}

		configuration.urls = fileLines
	}

	if url != "" {
		configuration.urls = append(configuration.urls, url)
	}

	if postDataFilePath != "" {
		configuration.method = "POST"
		data, err := os.ReadFile(postDataFilePath)
		if err != nil {
			log.Fatalf("Error in ioutil.ReadFile for file path: %s Error: %v", postDataFilePath, err)
		}
		configuration.postData = data
	}

	if postBody != "" {
		configuration.method = "POST"
		configuration.postData = []byte(postBody)
	}

	if contentType != "" {
		configuration.contentType = contentType
	}

	configuration.myClient.ReadTimeout = time.Duration(readTimeout) * time.Millisecond
	configuration.myClient.WriteTimeout = time.Duration(writeTimeout) * time.Millisecond
	configuration.myClient.MaxConnsPerHost = clients

	configuration.myClient.Dial = MyDialer()

	return configuration
}

func MyDialer() func(address string) (conn net.Conn, err error) {
	return func(address string) (net.Conn, error) {
		conn, err := net.Dial("tcp", address)
		if err != nil {
			return nil, err
		}

		myConn := &MyConn{Conn: conn}

		return myConn, nil
	}
}

func client(configuration configuration, result *Result, done *sync.WaitGroup) {
	// either perform requests until request count is reached or wait for timeout to kick in
	for result.requests < configuration.requestCount || configuration.requestsDuration != -1 {
		for _, tmpUrl := range configuration.urls {

			req := fasthttp.AcquireRequest()

			req.SetRequestURI(tmpUrl)
			req.Header.SetMethodBytes([]byte(configuration.method))

			if configuration.keepAlive {
				req.Header.Set("Connection", "keep-alive")
			} else {
				req.Header.Set("Connection", "close")
			}

			if len(configuration.authHeader) > 0 {
				req.Header.Set("Authorization", configuration.authHeader)
			}

			if len(configuration.contentType) > 0 {
				req.Header.SetContentType(configuration.contentType)
			}

			req.SetBody(configuration.postData)

			resp := fasthttp.AcquireResponse()
			err := configuration.myClient.Do(req, resp)
			statusCode := resp.StatusCode()
			result.requests++
			fasthttp.ReleaseRequest(req)
			fasthttp.ReleaseResponse(resp)

			if err != nil {
				result.networkFailed++
				continue
			}

			// check for any success status code
			if statusCode >= 200 && statusCode <= 226 {
				result.success++
			} else {
				result.badFailed++
			}
		}
	}

	done.Done()
}

func main() {
	startTime := time.Now()
	var done sync.WaitGroup
	results := make(map[int]*Result)

	configuration := newConfiguration()
	goMaxProcs := os.Getenv("GOMAXPROCS")
	if goMaxProcs == "" {
		runtime.GOMAXPROCS(runtime.NumCPU())
	}

	// interupt and print results on ctr+c
	Interrupted := make(chan os.Signal, 1)
	signal.Notify(Interrupted, os.Interrupt)

	// register timeout
	timeout := make(chan bool, 1)
	if configuration.requestsDuration != -1 {
		go func() {
			time.Sleep(time.Duration(configuration.requestsDuration) * time.Second)
			timeout <- true
		}()
	}

	go func() {
		select {
		case <-Interrupted:
		case <-timeout:
		}
		printResults(results, startTime)
		os.Exit(0)
	}()

	fmt.Printf("Dispatching %d clients\n", clients)

	done.Add(clients)
	for i := 0; i < clients; i++ {
		result := &Result{}
		results[i] = result
		go client(configuration, result, &done)

	}
	fmt.Println("Waiting for results...")
	done.Wait()
	printResults(results, startTime)
}
